charmhelpers/contrib/openstack/templating.py

```diff
@@ -238,29 +260,48 @@ def _get_template(self, template):
         log('Loaded template from %s' % template.filename, level=INFO)
         return template
 
+    def _get_template_from_string(self, ostmpl):
+        self._get_tmpl_env()
+        template = self._tmpl_env.from_string(ostmpl.config_template)
+        log('Loaded a template from a string for %s' % (
+            ostmpl.config_file),
+            level=INFO)
+        return template
+
     def render(self, config_file):
         if config_file not in self.templates:
             log('Config not registered: %s' % config_file, level=ERROR)
             raise OSConfigException
-        ctxt = self.templates[config_file].context()
-
-        _tmpl = os.path.basename(config_file)
-        try:
-            template = self._get_template(_tmpl)
-        except exceptions.TemplateNotFound:
-            # if no template is found with basename, try looking for it
-            # using a munged full path, eg:
-            #   /etc/apache2/apache2.conf -> etc_apache2_apache2.conf
-            _tmpl = '_'.join(config_file.split('/')[1:])
+
+        ostmpl = self.templates[config_file]
+        ctxt = ostmpl.context()
+
+        if ostmpl.from_string:
+            template = self._get_template_from_string(ostmpl)
+            log('Rendering from a string template: %s' % config_file,
+                level=INFO)
+        else:
+            _tmpl = os.path.basename(config_file)
             try:
                 template = self._get_template(_tmpl)
-            except exceptions.TemplateNotFound as e:
-                log('Could not load template from %s by %s or %s.' %
-                    (self.templates_dir, os.path.basename(config_file), _tmpl),
-                    level=ERROR)
-                raise e
-
-        log('Rendering from template: %s' % _tmpl, level=INFO)
+            except exceptions.TemplateNotFound:
+                # if no template is found with basename, try looking
+                # for it using a munged full path, eg:
+                # /etc/apache2/apache2.conf -> etc_apache2_apache2.conf
+                _tmpl = '_'.join(config_file.split('/')[1:])
```

Comment:
This assumes that the config_file always starts with a `/`.  Is that the case?  Can we stick an assert / if statement to verify as otherwise this may do odd things.

Reply:
Merely re-formatted this part to be included in the if statement branch. I think if changes are needed here it should be in a separate PR.

Reply:
Fair enough, ignore this one.